### PR TITLE
Add Azure Data Lake Gen 1 support for UFS

### PR DESCRIPTION
**Pull request by bastrik**
Adds ADL Gen1 UFS extension as a HDFS extension.

Attn @apc999 @madanadit

### CODE REVIEW

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      }
+      if (PropertyKey.Template.UNDERFS_AZURE_CLIENT_SECRET.matches(key)) {
+        adlConf.set(key, value);
+      }
+      if (PropertyKey.Template.UNDERFS_AZURE_REFRESH_URL.matches(key)) {
+        adlConf.set(key, value);
+      }
+    }
+    adlConf.set("fs.adl.oauth2.access.token.provider.type", "ClientCredential");
+//    adlConf.set("fs.adl.account.ruizhbigcacheadlsgen1.oauth2.access.token.provider.type", "ClientCredential");
```
Comment (apc999) [inline underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java:74]: plz remove commented lines if this is not meant to be used

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystemFactory.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  @Override
+  public UnderFileSystem create(String path, UnderFileSystemConfiguration conf) {
+    Preconditions.checkArgument(path != null, "path may not be null");
+    return AdlUnderFileSystem.createInstance(new AlluxioURI(path), conf);
+  }
+
+  @Override
+  public boolean supportsPath(String path) {
+    return path != null
+            && (path.startsWith(AdlUnderFileSystem.SCHEME_SECURE)
```
Comment (apc999) [inline underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystemFactory.java:43]: 4-space indentation for wrapped lines. 
see https://docs.alluxio.io/os/user/stable/en/contributor/Code-Conventions.html#javadoc-style

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  @Override
+  public UfsStatus getStatus(String path) throws IOException {
+    UfsStatus status = super.getStatus(path);
+    if (status instanceof UfsFileStatus) {
+      // adl is backed by an object store but always claims its block size to be 512MB.
+      // reset the block size in UfsFileStatus according to getBlockSizeByte
+      return new UfsFileStatus(path,
+              ((UfsFileStatus) status).getContentHash(),
```
Comment (apc999) [inline underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java:121]: indentation

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+  /**
+   * Factory method to construct a new Adl {@link UnderFileSystem}.
+   *
+   * @param uri the {@link AlluxioURI} for this UFS
+   * @param conf the configuration for this UFS
+   * @return a new Adl {@link UnderFileSystem} instance
+   */
+  public static AdlUnderFileSystem createInstance(AlluxioURI uri,
+                                                  UnderFileSystemConfiguration conf) {
```
Comment (apc999) [inline underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java:86]: indentation, https://docs.alluxio.io/os/user/stable/en/contributor/Code-Conventions.html#javadoc-style

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An {@link UnderFileSystem} uses the Microsoft Azure Blob Storage.
```
Comment (madanadit) [inline underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java:39]: is this comment stale?

**Review by madanadit — COMMENTED** [review]
Thanks @bastrik for your contribution. Assuming ADLS gen 2, also supports the same hdfs compatible interface, this ufs should work with both gen1 and gen2?

Also, please plan to run tests using https://docs.alluxio.io/os/user/stable/en/operation/User-CLI.html#runufstests, if not already. These tests verify the expected contract b/w Alluxio and the under storage. Thank you.

**Comment by bastrik** [comment]
> Can ADLS Gen1 replace wasb UFS? If yes, can we remove wasb UFS?

ADLS Gen1 and wasb has different implementations and actually uses different modules (hadoop-azure-datalake vs. hadoop-azure), so both of these are needed.

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java (hunk 5 above)
Comment (bastrik) [reply underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java]: fixed

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java (hunk 1 above)
Comment (bastrik) [reply underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java]: fixed

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java (hunk 4 above)
Comment (bastrik) [reply underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java]: fixed

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java (hunk 3 above)
Comment (bastrik) [reply underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystem.java]: fixed

File: underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystemFactory.java (hunk 2 above)
Comment (bastrik) [reply underfs/adl/src/main/java/alluxio/underfs/adl/AdlUnderFileSystemFactory.java]: fixed

**Review by madanadit — APPROVED** [review]
Thank you

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
@bastrik can you follow instruction https://github.com/Alluxio/alluxio/pull/12047#issuecomment-683093381 to link your email correctly , so we could keep your credits in release notes? thanks

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Comment by apc999** [comment]
alluxio-bot, cherry-pick this to branch-2.3-fuse please

**Pull request merged**